console implementation according to standard

lal12 opened the pull request:
Implementing console according to standard.

Still a bunch of test cases missing, but otherwise complete.

Any thoughts?

lal12 commented:
E.g. would be nice to expose the createConsole function for customization.

saghul commented on src/bundles/c/stdlib/assert.c:
```diff
@@ -361,9 +361,9 @@ const uint8_t tjs__assert[8835] = {
  0x75, 0x72, 0x6e, 0x20, 0x61, 0x20, 0x21, 0x3d,
```
(random comment placement).

What did you use to regenerate the code? You only changed a polyfill, so stdlib shouldn't be regenerated I think.

lal12 replied:
`make js`. Yes, you're right. I think I was on an unclean checkout (on another notebook) when force pushing it last night.

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
+    if (!logger) {
+        logger = function Logger(logLevel, args) {
+            if (args.length === 1) {
+                _printer(logLevel, args[0]);
```
Why is the single argument not formatted here?

lal12 replied:
See spec: https://console.spec.whatwg.org/#logger

saghul replied:
👍

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
-function hasOwnProperty(obj, v) {
-    if (obj === null  || typeof obj === 'undefined') {
-        return false;
+    if (!formatter) {
```
can we use defaults instead please?

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
 
-    return Object.prototype.hasOwnProperty.call(obj, v);
-}
+    if (!inspect) {
```
Ditto.

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
+        },
+        clear() {
+            groupCount = 0;
+            clear();
```
This reads weird, can we rename the auxiliary function?

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
-    info(...args) {
-        print(...args);
-    }
+            for (let ci=0;ci<rows[0].length;ci++) {
```
style: please add spacing

lal12 replied:
Not sure where exactly you meant. Maybe got lost during force-push. But eslint doesn't complain.

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
+            _printer('count', label + ': ' + count);
+        },
+        countReset(label = 'default') {
+            if (!countMap.delete(label)) {
```
When is `count` decreased?

saghul replied:
Sorry, my bad, I misunderstood what the API did 😅

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
+    })
+});
+
+Object.defineProperty(window, 'createConsole', {
```
Let's attach this to the `tjs` global instead, since it's non-standard.

saghul commented on tests/test-console.js:
```diff
@@ -0,0 +1,80 @@
+
+];
+
+(async ()=>{
```
No need for the immediate function.

saghul commented on tests/test-console.js:
```diff
@@ -0,0 +1,80 @@
+		const status = await proc.wait();
+		const buf = new Uint8Array(4096);
+		const nread = await proc.stdout.read(buf);
+		const stdout = new TextDecoder().decode(buf.subarray(0, nread));
```
Create and reuse a decoder.

saghul commented on tests/test-console.js:
```diff
@@ -0,0 +1,80 @@
+
+		const {code, stdout, stderr} = await runTest(jscode);
+		if(code !== 0){
+			console.error(`Test failed with code ${code}: ${jscode}`);
```
This should actually fail the test.

saghul reviewed (changes requested):
Great work! Left some comments.

saghul commented on src/js/polyfills/console.js:
```diff
@@ -1,258 +1,316 @@
+            _printer('count', label + ': ' + count);
+        },
+        countReset(label = 'default') {
+            if (!countMap.delete(label)) {
```
According to the spec a reset sets it to 0, it doesn't delete it. Is trhere a practical difference?

lal12 replied:
Yep, there is. When it is set to 0 you won't get a warning the next time when calling `.countReset()` again.

Though this is a weird case and nodejs does it the way I do. And I think it much better otherwise you never could clean up a counter and it would leak stuff.

```
> console.count('a')
a: 1
undefined
> console.countReset('a')
undefined
> console.countReset('a')
undefined
> (node:18291) Warning: Count for 'a' does not exist
(Use `node --trace-warnings ...` to show where the warning was created)
```

saghul replied:
Very subtle, let's keep it as is not to have an ever growing map indeed.

lal12 commented:
Added tests from wpt and added some changes to fulfill those and match some nodejs behavior. In some places wpt tests seemed to be wrong, and neither browser nor nodejs acted accordingly (but differently from each other), there I used nodejs behavior.

JS console standards seems to be oddly wild :laughing: 

Diverging from spec are the following things:

- countReset is resetting instead setting to 0. MDN leaves it open, NodeJS does this too, Chrome+FF follow spec, though FF also prints `label: 0`
- added options argument to Logger, to allow for some specific warning handling to better match nodejs behavior

I also copied the code from util package and moved it here to add missing features as defined in console spec. Though alternatively we could extend the alteady existing util patch.

saghul commented:
This looks great! Are you going to be adding anything else, or should we merge?

lal12 commented:
> This looks great! Are you going to be adding anything else, or should we merge?

Just pushed the types for createConsole. I have nothing else to add. The only thing I was wondering if there is a better way for registering `createConsole` (see `console.js:327`)

saghul commented:
> The only thing I was wondering if there is a better way for registering `createConsole` (see `console.js:327`)

It looks odd indeed. I'll take a look.

saghul commented:
Excellent work @lal12 !
